apps/axistream/src/axi_stream_reader.hh

```diff
@@ -0,0 +1,403 @@
+    typename std::enable_if<T::width == H::width, T>::type read_after(const H& header) {
+        HLS_PRAGMA(HLS inline);
+        return T::from_raw(H::to_raw(header));
+    }
+
+
+    /*
+     * Reads the specified number of byes from the stream.
+     */
+    template <typename LengthPolicy, int MsgWidth>
+    typename LengthPolicy::template Result<MsgWidth> read_raw() {
+        HLS_PRAGMA(HLS inline);
+        // The size of the working buffer.
+        // This handles the worst-case where we have DataWidth - 1 bytes in buf_,
+        // and so the last byte of the message data will be at byte index DataWidth - 1 + MsgWidth - 1
+        // (we will need to store DataWidth - 1 + MsgWidth bytes).
+        // The work size, in the last data beat read, the high byte will not be part of this message,
+        // but we still must store it in the work buffer.
+        // Thus the work buffer neesd to be DataWidth + MsgWidth bytes.
+        const int work_size = util::BITS_PER_BYTE * (DataWidth + MsgWidth);
+        // Maximum number of beats required to get the whole message = ceil(MsgWidth / DataWidth)
+        // It is possible the last beat isn't needed depending on how much is stored in buf_
+        // from the last read
+        const int max_beats = (MsgWidth + DataWidth - 1) / DataWidth;
+        // Maximum number of new bytes we might receive
+        const int max_received = DataWidth * max_beats;
+        // Type which can store the number of bytes to shift the final result by
+        // At most, all the low DataWidth bytes need to be shifted
+        using Shift = util::uint_t<DataWidth * util::BITS_PER_BYTE>;
+
+        // Buf will be zero extended since work is MsgWidth longer than buf_
+        ap_uint<work_size> work = buf_;
+        // Represents the number of bytes we have so far
+		using Received = util::uint_t<max_received>;
+        Received received = len_;
+        Received received_upper_bound = len_;
+        // Represents the index in work where the last byte of the message will be written
+        util::state_t<work_size> last_index = DataWidth - len_ + MsgWidth - 1;
+
+        for (int i = 0; i < max_beats; i++) {
+            HLS_PRAGMA(HLS unroll);
+            // Not including this if statement in the loop guard allows Vivado to generate better
+            // code because then the loop has a fixed number of iterations.
+            // Only check received_upper_bound - not the actual received!
+            // This is because received_upper_bound == received _unless_ last_ is set --
+            // each data beat contains DataWidth bytes unless last_ is set
+            // Using received_upper_bound here instead of received helps the HLS tool
+            // because then it is evident that the loop only exits on last_,
+            // and not the uncontrained output of compute_received_delta.
+			if (received_upper_bound >= Received(MsgWidth) || !LengthPolicy::more_available(last_)) {
```

Comment:
The `LengthPolicy` enables `Checked` and `Unchecked` reads. `Checked` reads respect keep & last, and return both the data read, and the number of bytes actually read.